src/io/flutter/actions/OpenFlutterNewsSidePanelAction.java

```diff
@@ -0,0 +1,166 @@
+
+    FlutterNewsToolWindowListener.fireOpenEvent(project, isAutoOpened);
+    openWindow(FlutterNewsBundleCreator.BUNDLE_ID, project);
+
+    // Only register a new listener if there isn't already one, to avoid multiple OPEN/CLOSE events
+    myProjectToListenerMap.computeIfAbsent(project, this::newFlutterNewsToolWindowListener);
+  }
+
+  @NotNull
+  private OpenFlutterNewsSidePanelAction.FlutterNewsToolWindowListener newFlutterNewsToolWindowListener(@NotNull Project project) {
+    FlutterNewsToolWindowListener listener = new FlutterNewsToolWindowListener(project, myProjectToListenerMap);
+    project.getMessageBus().connect(project).subscribe(ToolWindowManagerListener.TOPIC, listener);
+    return listener;
+  }
+
+  static class FlutterNewsToolWindowListener implements ToolWindowManagerListener {
+    @NotNull private final Project myProject;
+    @NotNull Map<Project, FlutterNewsToolWindowListener> myProjectToListenerMap;
+    private boolean isOpen;
+
+    private FlutterNewsToolWindowListener(@NotNull Project project,
+                                          @NotNull Map<Project, FlutterNewsToolWindowListener> projectToListenerMap) {
+      myProject = project;
+      myProjectToListenerMap = projectToListenerMap;
+      isOpen = true; // Start off as opened so we don't fire an extra opened event
+
+      // Need an additional listener for project close, because the below invokeLater isn't fired in time before closing
+      project.getMessageBus().connect(project).subscribe(ProjectManager.TOPIC, new ProjectManagerListener() {
+        @Override
+        public void projectClosed(@NotNull Project project) {
+          if (!project.equals(myProject)) {
+            return;
+          }
+          if (isOpen) {
+            fireClosedEvent(myProject);
+            isOpen = false;
+          }
+          myProjectToListenerMap.remove(project);
+        }
+      });
+    }
+
+    @Override
+    public void toolWindowRegistered(@NotNull String id) {
+    }
+
+    @Override
+    public void toolWindowUnregistered(@NotNull String id, @NotNull ToolWindow toolWindow) {
+      if (id.equals(OpenAssistSidePanelAction.TOOL_WINDOW_TITLE)) {
+        myProjectToListenerMap.remove(myProject);
+      }
+    }
+
+    /**
+     * Fire metrics and update the actual state after a state change is received.
+     * The logic is wrapped in invokeLater because dragging and dropping the StripeButton temporarily
+     * hides and then shows the window. Otherwise, the handler would think the window was closed,
+     * even though it was only dragged.
+     */
+    @Override
+    public void stateChanged(@NotNull ToolWindowManager toolWindowManager) {
+      ApplicationManager.getApplication().invokeLater(() -> {
+        if (myProject.isDisposed()) {
+          myProjectToListenerMap.remove(myProject);
+          return;
+        }
+
+        ToolWindow window = ToolWindowManager.getInstance(myProject).getToolWindow(OpenAssistSidePanelAction.TOOL_WINDOW_TITLE);
```

Comment:
Is the argument `toolWindowManager` the same as what's retrieved here?

Reply:
Not necessarily. The tool window manager is project-specific. This could have been triggered by a different project. It might make sense to check that it is same and do nothing if not.